kvittering/src/main/java/no/difi/meldingsutveksling/kvittering/SBDReceiptFactory.java

```diff
@@ -0,0 +1,140 @@
+                                                            DocumentType documentType,
+                                                            ReceiptStatus status) {
+        String standard;
+        try {
+            standard = serviceRegistryLookup.getDocumentIdentifier(SRParameter.builder(sbd.getSenderIdentifier())
+                    .conversationId(sbd.getConversationId()).build(), Process.EINNSYN_RESPONSE, documentType);
+        } catch (ServiceRegistryLookupException e) {
+            log.error("Error looking up service record for {}", sbd.getSenderIdentifier(), e);
+            throw new MeldingsUtvekslingRuntimeException(e);
+        }
+        StatusMessage statusMessage = new StatusMessage(status);
+
+        return new StandardBusinessDocument()
+                .setStandardBusinessDocumentHeader(new StandardBusinessDocumentHeader.Builder()
+                        .from(Organisasjonsnummer.from(sbd.getReceiverIdentifier()))
+                        .to(Organisasjonsnummer.from(sbd.getSenderIdentifier()))
+                        .relatedToConversationId(sbd.getConversationId())
+                        .process(Process.EINNSYN_RESPONSE)
+                        .standard(standard)
+                        .type(documentType)
+                        .build())
+                .setAny(statusMessage);
+    }
+
+    private StandardBusinessDocument signAndWrapDocument(MessageInfo messageInfo, IntegrasjonspunktNokkel keyInfo, Kvittering kvittering) {
+
+        StandardBusinessDocument unsignedReceipt = new StandardBusinessDocument();
+        StandardBusinessDocumentHeader header = new StandardBusinessDocumentHeader.Builder()
+                // sender of the receipt is the receiver of the message
+                .from(new Organisasjonsnummer(messageInfo.getReceiverOrgNumber()))
+                .to(new Organisasjonsnummer(messageInfo.getSenderOrgNumber()))
+                .relatedToConversationId(messageInfo.getConversationId())
+                .relatedToJournalPostId(messageInfo.getJournalPostId())
+                .process(Process.LEGACY)
+                .standard(Standard.LEGACY.getValue())
+                .type(DocumentType.BESTEDU_KVITTERING)
+                .build();
+
+        unsignedReceipt.setStandardBusinessDocumentHeader(header);
+        unsignedReceipt.setAny(new ObjectFactory().createKvittering(kvittering));
+
+        org.w3c.dom.Document xmlDoc = toXMLDocument(unsignedReceipt);
+        org.w3c.dom.Document signedXmlDoc = DocumentSigner.sign(xmlDoc, keyInfo);
+        try {
+            if (!DocumentValidator.validate(signedXmlDoc)) {
+                throw new MeldingsUtvekslingRuntimeException("created non validating document");
+            }
+        } catch (XPathExpressionException | XMLSecurityException e) {
+            throw new RuntimeException("Could not validate signature that we used to sign the document", e);
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar-app01.difi.local:9000/coding_rules#rule_key=squid%3AS00112)